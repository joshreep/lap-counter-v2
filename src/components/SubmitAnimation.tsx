import { LottieComponentProps } from 'lottie-react'
import { FC, useCallback } from 'react'
import LoadingAnimation from './LoadingAnimation'
import CheckAnimation from './CheckAnimation'

export enum SubmissionState {
  Idle,
  Pending,
  Complete,
  Error,
}

interface SubmitAnimationProps extends Omit<LottieComponentProps, 'animationData'> {
  submissionState: SubmissionState
  onAnimationFinish: NonNullable<LottieComponentProps['onComplete']>
}

const SubmitAnimation: FC<SubmitAnimationProps> = (props) => {
  const { onAnimationFinish, submissionState } = props
  const onCheckAnimationFinished = useCallback(() => {
    setTimeout(() => {
      onAnimationFinish(undefined)
    }, 150)
  }, [onAnimationFinish])

  if (submissionState === SubmissionState.Pending) return <LoadingAnimation />

  if (submissionState === SubmissionState.Complete)
    return <CheckAnimation onComplete={onCheckAnimationFinished} />

  return null
}

export default SubmitAnimation
