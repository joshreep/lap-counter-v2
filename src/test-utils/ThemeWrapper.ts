import MatchMediaMock from 'jest-matchmedia-mock'

type Theme = 'light' | 'dark'

const themes: Theme[] = ['light', 'dark']

export function forEachTheme(callback: (theme: Theme) => void) {
  themes.forEach((theme) => {
    const matchMedia = new MatchMediaMock()
    matchMedia.useMediaQuery(`(prefers-color-scheme: ${theme})`)

    callback(theme)

    matchMedia.clear()
  })
}
