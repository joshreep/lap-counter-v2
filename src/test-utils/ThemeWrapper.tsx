import SCDarkTheme from '@/theme/dark-theme'
import SCLightTheme from '@/theme/light-theme'
import {
  DarkTheme as RNDarkTheme,
  DefaultTheme as RNDefaultTheme,
  ThemeProvider as RNThemeProvider,
} from '@react-navigation/native'
import { FC, PropsWithChildren } from 'react'
import { ColorSchemeName } from 'react-native'
import { ThemeProvider as SCThemeProvider } from 'styled-components/native'

export const themes: ColorSchemeName[] = ['light', 'dark']

export function getThemeWrapper(theme: ColorSchemeName) {
  const Wrapper: FC<PropsWithChildren> = ({ children }) => {
    return (
      <RNThemeProvider value={theme === 'dark' ? RNDarkTheme : RNDefaultTheme}>
        <SCThemeProvider theme={theme === 'dark' ? SCDarkTheme : SCLightTheme}>{children}</SCThemeProvider>
      </RNThemeProvider>
    )
  }

  return Wrapper
}

type ForEachThemeCallback = (theme: ColorSchemeName, wrapper: FC<PropsWithChildren>) => void

export function forEachTheme(callback: ForEachThemeCallback) {
  themes.forEach((theme) => {
    callback(theme, getThemeWrapper(theme))
  })
}
