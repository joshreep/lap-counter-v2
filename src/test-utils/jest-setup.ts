import '@testing-library/react-native/extend-expect'
import '@testing-library/jest-native/extend-expect'

jest.mock('react-native-reanimated', () => null, { virtual: true })
jest.mock('@testing-library/jest-native/extend-expect', () => null, { virtual: true })
