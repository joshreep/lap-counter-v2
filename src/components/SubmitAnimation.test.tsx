import { forEachTheme } from '@/test-utils/ThemeWrapper'
import { render } from '@testing-library/react-native'
import SubmitAnimation, { SubmissionState } from './SubmitAnimation'
import { FC, PropsWithChildren } from 'react'

type SetupProps = {
  submissionState: SubmissionState
  wrapper: FC<PropsWithChildren>
}

function setup({ submissionState, wrapper }: SetupProps) {
  jest.useFakeTimers()
  const onAnimationFinishMock = jest.fn()
  const container = render(
    <SubmitAnimation submissionState={submissionState} onAnimationFinish={onAnimationFinishMock} />,
    { wrapper },
  )

  return { ...container, onAnimationFinishMock }
}

forEachTheme((theme, wrapper) => {
  describe(`${theme} theme`, () => {
    test('should render null when submissionState is Idle', () => {
      const container = setup({ submissionState: SubmissionState.Idle, wrapper })
      expect(container.onAnimationFinishMock).not.toHaveBeenCalled()
      expect(container.toJSON()).toBeNull()
    })

    test('should render null when submissionState is Error', () => {
      const container = setup({ submissionState: SubmissionState.Error, wrapper })
      expect(container.onAnimationFinishMock).not.toHaveBeenCalled()
      expect(container.toJSON()).toBeNull()
    })

    test('should render LoadingAnimation when submissionState is Pending', () => {
      const container = setup({ submissionState: SubmissionState.Pending, wrapper })
      expect(container.onAnimationFinishMock).not.toHaveBeenCalled()
      expect(container.toJSON()).toMatchSnapshot()
    })

    test('should render CheckAnimation when submissionState is Complete', () => {
      jest.mock('lottie-react-native')
      const container = setup({ submissionState: SubmissionState.Complete, wrapper })
      expect(container.toJSON()).toMatchSnapshot()
      jest.runAllTimers()
      expect(container.onAnimationFinishMock).toHaveBeenCalled()
    })
  })
})
