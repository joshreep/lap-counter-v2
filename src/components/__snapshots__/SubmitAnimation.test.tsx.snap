// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark theme should render CheckAnimation when submissionState is Complete 1`] = `
<div
  autoplay=""
  class="w-3xs h-3xs"
>
  <svg
    height="151"
    preserveAspectRatio="xMidYMid meet"
    style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
    viewBox="0 0 153 151"
    width="153"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <clippath
        id="__lottie_element_54"
      >
        <rect
          height="151"
          width="153"
          x="0"
          y="0"
        />
      </clippath>
      <clippath
        id="__lottie_element_56"
      >
        <path
          d="M0,0 L135,0 L135,128 L0,128z"
        />
      </clippath>
    </defs>
    <g
      clip-path="url(#__lottie_element_54)"
    >
      <g
        clip-path="url(#__lottie_element_56)"
        opacity="0.00029343923051978793"
        style="display: block;"
        transform="matrix(1.1189650297164917,0,0,1.1189650297164917,0.8993638753890991,3.477238178253174)"
      >
        <g
          style="display: none;"
        >
          <g>
            <path />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.0002934392250608653,0,0,0.0002934392250608653,67.47222137451172,63.61372375488281)"
        >
          <g
            opacity="1"
            transform="matrix(1,0,0,1,-5,-4)"
          >
            <path
              d=" M31,3.5 C45.62535095214844,3.5 57.5,15.374650001525879 57.5,30 C57.5,44.62535095214844 45.62535095214844,56.5 31,56.5 C16.374649047851562,56.5 4.5,44.62535095214844 4.5,30 C4.5,15.374650001525879 16.374649047851562,3.5 31,3.5z"
              fill="rgb(35,178,109)"
              fill-opacity="1"
            />
            <path
              d=" M31,3.5 C45.62535095214844,3.5 57.5,15.374650001525879 57.5,30 C57.5,44.62535095214844 45.62535095214844,56.5 31,56.5 C16.374649047851562,56.5 4.5,44.62535095214844 4.5,30 C4.5,15.374650001525879 16.374649047851562,3.5 31,3.5z"
              fill-opacity="0"
              stroke="rgb(0,0,0)"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
              stroke-opacity="0"
              stroke-width="0"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`dark theme should render LoadingAnimation when submissionState is Pending 1`] = `
<div
  autoplay=""
  class="w-3xs h-3xs"
>
  <svg
    height="300"
    preserveAspectRatio="xMidYMid meet"
    style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
    viewBox="0 0 300 300"
    width="300"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <clippath
        id="__lottie_element_46"
      >
        <rect
          height="300"
          width="300"
          x="0"
          y="0"
        />
      </clippath>
    </defs>
    <g
      clip-path="url(#__lottie_element_46)"
    >
      <g
        opacity="0.3"
        style="display: block;"
        transform="matrix(0.30000001192092896,0,0,0.30000001192092896,150,150)"
      >
        <g
          opacity="1"
          transform="matrix(1,0,0,1,0,0)"
        >
          <path
            d=" M0,-150 C82.78500366210938,-150 150,-82.78500366210938 150,0 C150,82.78500366210938 82.78500366210938,150 0,150 C-82.78500366210938,150 -150,82.78500366210938 -150,0 C-150,-82.78500366210938 -82.78500366210938,-150 0,-150z"
            fill-opacity="0"
            stroke="rgb(170,215,235)"
            stroke-linecap="butt"
            stroke-linejoin="miter"
            stroke-miterlimit="4"
            stroke-opacity="1"
            stroke-width="50"
          />
        </g>
      </g>
      <g
        opacity="1"
        style="display: block;"
        transform="matrix(0.30000001192092896,0.000031415926059708,-0.000031415926059708,0.30000001192092896,150,150)"
      >
        <g
          opacity="1"
          transform="matrix(1,0,0,1,0,0)"
        >
          <path
            d=" M0,-150 C3.1640000343322754,-150 6.303999900817871,-149.90199279785156 9.418999671936035,-149.70799255371094"
            fill-opacity="0"
            stroke="rgb(58,169,220)"
            stroke-linecap="round"
            stroke-linejoin="miter"
            stroke-miterlimit="4"
            stroke-opacity="1"
            stroke-width="50"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`light theme should render CheckAnimation when submissionState is Complete 1`] = `
<div
  autoplay=""
  class="w-3xs h-3xs"
>
  <svg
    height="151"
    preserveAspectRatio="xMidYMid meet"
    style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
    viewBox="0 0 153 151"
    width="153"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <clippath
        id="__lottie_element_10"
      >
        <rect
          height="151"
          width="153"
          x="0"
          y="0"
        />
      </clippath>
      <clippath
        id="__lottie_element_12"
      >
        <path
          d="M0,0 L135,0 L135,128 L0,128z"
        />
      </clippath>
    </defs>
    <g
      clip-path="url(#__lottie_element_10)"
    >
      <g
        clip-path="url(#__lottie_element_12)"
        opacity="0.00029343923051978793"
        style="display: block;"
        transform="matrix(1.1189650297164917,0,0,1.1189650297164917,0.8993638753890991,3.477238178253174)"
      >
        <g
          style="display: none;"
        >
          <g>
            <path />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path
              fill-rule="evenodd"
            />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.0002934392250608653,0,0,0.0002934392250608653,67.47222137451172,63.61372375488281)"
        >
          <g
            opacity="1"
            transform="matrix(1,0,0,1,-5,-4)"
          >
            <path
              d=" M31,3.5 C45.62535095214844,3.5 57.5,15.374650001525879 57.5,30 C57.5,44.62535095214844 45.62535095214844,56.5 31,56.5 C16.374649047851562,56.5 4.5,44.62535095214844 4.5,30 C4.5,15.374650001525879 16.374649047851562,3.5 31,3.5z"
              fill="rgb(35,178,109)"
              fill-opacity="1"
            />
            <path
              d=" M31,3.5 C45.62535095214844,3.5 57.5,15.374650001525879 57.5,30 C57.5,44.62535095214844 45.62535095214844,56.5 31,56.5 C16.374649047851562,56.5 4.5,44.62535095214844 4.5,30 C4.5,15.374650001525879 16.374649047851562,3.5 31,3.5z"
              fill-opacity="0"
              stroke="rgb(0,0,0)"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
              stroke-opacity="0"
              stroke-width="0"
            />
          </g>
        </g>
        <g
          style="display: none;"
        >
          <g>
            <path />
            <path
              fill-opacity="0"
              stroke-linecap="square"
              stroke-linejoin="miter"
              stroke-miterlimit="1"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`light theme should render LoadingAnimation when submissionState is Pending 1`] = `
<div
  autoplay=""
  class="w-3xs h-3xs"
>
  <svg
    height="300"
    preserveAspectRatio="xMidYMid meet"
    style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
    viewBox="0 0 300 300"
    width="300"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <clippath
        id="__lottie_element_2"
      >
        <rect
          height="300"
          width="300"
          x="0"
          y="0"
        />
      </clippath>
    </defs>
    <g
      clip-path="url(#__lottie_element_2)"
    >
      <g
        opacity="0.3"
        style="display: block;"
        transform="matrix(0.30000001192092896,0,0,0.30000001192092896,150,150)"
      >
        <g
          opacity="1"
          transform="matrix(1,0,0,1,0,0)"
        >
          <path
            d=" M0,-150 C82.78500366210938,-150 150,-82.78500366210938 150,0 C150,82.78500366210938 82.78500366210938,150 0,150 C-82.78500366210938,150 -150,82.78500366210938 -150,0 C-150,-82.78500366210938 -82.78500366210938,-150 0,-150z"
            fill-opacity="0"
            stroke="rgb(170,215,235)"
            stroke-linecap="butt"
            stroke-linejoin="miter"
            stroke-miterlimit="4"
            stroke-opacity="1"
            stroke-width="50"
          />
        </g>
      </g>
      <g
        opacity="1"
        style="display: block;"
        transform="matrix(0.30000001192092896,0.000031415926059708,-0.000031415926059708,0.30000001192092896,150,150)"
      >
        <g
          opacity="1"
          transform="matrix(1,0,0,1,0,0)"
        >
          <path
            d=" M0,-150 C3.1640000343322754,-150 6.303999900817871,-149.90199279785156 9.418999671936035,-149.70799255371094"
            fill-opacity="0"
            stroke="rgb(58,169,220)"
            stroke-linecap="round"
            stroke-linejoin="miter"
            stroke-miterlimit="4"
            stroke-opacity="1"
            stroke-width="50"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;
